File: frontend/src/api/uploads.ts
import type { ImageUploadResponse } from '../types';

const BASE_URL = import.meta.env.VITE_API_URL || 'http://localhost:8000/api/v1';

/**
 * Upload an image file to a course
 */
export async function uploadImage(
  courseId: string,
  file: File
): Promise<ImageUploadResponse> {
  const token = localStorage.getItem('token');

  const formData = new FormData();
  formData.append('file', file);

  const response = await fetch(`${BASE_URL}/uploads/images/${courseId}`, {
    method: 'POST',
    headers: {
      ...(token && { Authorization: `Bearer ${token}` }),
    },
    body: formData,
  });

  if (!response.ok) {
    const error = await response.json().catch(() => ({}));
    throw new Error(error.detail || 'Failed to upload image');
  }

  return response.json();
}

/**
 * Upload an image from a URL to a course
 */
export async function uploadImageFromUrl(
  courseId: string,
  url: string
): Promise<ImageUploadResponse> {
  const token = localStorage.getItem('token');

  const response = await fetch(`${BASE_URL}/uploads/images/${courseId}/from-url`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      ...(token && { Authorization: `Bearer ${token}` }),
    },
    body: JSON.stringify({ url }),
  });

  if (!response.ok) {
    const error = await response.json().catch(() => ({}));
    throw new Error(error.detail || 'Failed to upload image from URL');
  }

  return response.json();
}

/**
 * Get the full URL for a course image
 */
export function getImageUrl(courseId: string, filename: string): string {
  return `${BASE_URL}/uploads/courses/${courseId}/images/${filename}`;
}

export const uploadsApi = {
  uploadImage,
  uploadImageFromUrl,
  getImageUrl,
};

File: backend/api/routes/uploads.py
"""Upload API routes for images and files."""

from fastapi import APIRouter, Depends, UploadFile, File, HTTPException
from fastapi.responses import FileResponse
from pydantic import BaseModel

from dependencies import CurrentUser, StorageDep
from services.image_service import ImageService
from services.course_service import CourseService

router = APIRouter()


def get_image_service() -> ImageService:
    return ImageService()


def get_course_service(storage: StorageDep) -> CourseService:
    return CourseService(storage)


class ImageUploadResponse(BaseModel):
    """Response after image upload."""

    url: str
    filename: str


class ImageUrlRequest(BaseModel):
    """Request to download image from URL."""

    url: str


@router.post("/images/{course_id}", response_model=ImageUploadResponse)
async def upload_image(
    course_id: str,
    user: CurrentUser,
    file: UploadFile = File(...),
    image_service: ImageService = Depends(get_image_service),
    course_service: CourseService = Depends(get_course_service),
):
    """Upload an image file for a course.

    The user must be the author of the course.
    """
    # Verify user owns the course
    await course_service._get_editable_course(course_id, user.id)

    # Read file content
    content = await file.read()

    # Upload image
    filename = file.filename or "image.jpg"
    url = await image_service.upload_image(content, filename, course_id)

    return ImageUploadResponse(url=url, filename=filename)


@router.post("/images/{course_id}/from-url", response_model=ImageUploadResponse)
async def upload_image_from_url(
    course_id: str,
    data: ImageUrlRequest,
    user: CurrentUser,
    image_service: ImageService = Depends(get_image_service),
    course_service: CourseService = Depends(get_course_service),
):
    """Download an image from URL and store it for a course.

    The user must be the author of the course.
    """
    # Verify user owns the course
    await course_service._get_editable_course(course_id, user.id)

    # Download and store image
    url = await image_service.download_external_image(data.url, course_id)

    # Extract filename from URL
    from pathlib import Path

    filename = Path(url).name

    return ImageUploadResponse(url=url, filename=filename)


@router.get("/courses/{course_id}/images/{filename}")
async def get_image(
    course_id: str,
    filename: str,
    image_service: ImageService = Depends(get_image_service),
):
    """Serve an uploaded image.

    This endpoint is public to allow image display in markdown content.
    """
    file_path = image_service.get_image_path(course_id, filename)

    if not file_path:
        raise HTTPException(status_code=404, detail="Image not found")

    # Determine media type
    media_types = {
        ".png": "image/png",
        ".jpg": "image/jpeg",
        ".jpeg": "image/jpeg",
        ".gif": "image/gif",
        ".svg": "image/svg+xml",
        ".webp": "image/webp",
    }
    ext = file_path.suffix.lower()
    media_type = media_types.get(ext, "application/octet-stream")

    return FileResponse(file_path, media_type=media_type)
